sdk/compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -1474,6 +1490,80 @@ execDocTest opts scriptDar (ImportSource importSource) files =
       withDamlIdeState opts logger diagnosticsLogger $ \ideState ->
           docTest ideState files'
 
+data MultiPackageManifestEntry = MultiPackageManifestEntry
+  { packageName :: String
+  , packageVersion :: String
+  , packageDir :: FilePath
+  , packageSrc :: FilePath
+  , packageFileHashes :: Map.Map FilePath String
+  , packageHash :: String
+  , output :: FilePath
+  , packageDeps :: [String]
+  , darDeps :: [FilePath]
+  }
+  deriving (Show, Eq, Generic, ToJSON, FromJSON)
+
+-- Map can't be a bifunctor because of this `Ord` constraint, but its a shame such a function is defined in Data.Map
+bimapMap :: Ord k' => (k -> k') -> (v -> v') -> Map.Map k v -> Map.Map k' v'
+bimapMap keyMap valueMap = fmap valueMap . Map.mapKeys keyMap
+
+execGenerateMultiPackageManifest :: SdkVersion.Class.SdkVersioned => MultiPackageLocation -> GenerateMultiPackageManifestOutput -> Command
+execGenerateMultiPackageManifest multiPackageLocation outputLocation =
+    Command GenerateMultiPackageManifest Nothing effect
+  where
+    effect = do
+      mMultiPackageConfigProjectPath <- getMultiPackagePath multiPackageLocation
+      multiPackageConfigProjectPath <- case mMultiPackageConfigProjectPath of
+        Nothing -> do
+          hPutStrLn stderr "Couldn't find a multi-package.yaml at current or parent directory. Use --multi-package-path to specify its location."
+          exitFailure
+        Just path -> pure path
+      let multiPackageConfigPath = unwrapProjectPath multiPackageConfigProjectPath
+      withMultiPackageConfig multiPackageConfigProjectPath $ \multiPackageConfig -> do
+        configs <- forM (mpPackagePaths multiPackageConfig) $ \path -> do
+          config@BuildMultiPackageConfig {..} <- buildMultiPackageConfigFromDamlYaml path
+          darPath <- deriveDarPath path bmName bmVersion bmOutput
+          pure (path, darPath, config)
+
+        -- Map from dar path to config, for classifying data deps
+        let configMap = Map.fromList $ (\(_, darPath, config) -> (darPath, config)) <$> configs
+
+        manifestEntries <- forM configs $ \(packagePath, darPath, BuildMultiPackageConfig {..}) -> do
+          damlFilesMap <- bimapMap (bmSourceDaml </>) BSL.toStrict <$> getDamlFilesBuildMulti (hPutStrLn stderr) packagePath bmSourceDaml
```

Comment:
This function `getDamlFilesBuildMulti` will fail if a source file is specified instead of a directory - do we want to handle the file case as well?

Also, I noticed that the comment on `getDamlFilesBuildMulti` is wrong - it claims that an error will be raised, but afaict it only raises a warning.

Reply:
We can't fully handle the file case for the same reason we don't with multi-package caching, it requires parsing the file.
As you've noticed though, it will warn. We could include the hash of the file itself, but can't know the deps.
One option would be a hash of all daml files in the directory, giving a hash that would likely give false positives, but hopefully no false negatives? We'd need to include these files in the per file hash map though.